net/udp_any.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package net
+
+import (
+	"errors"
+	"net"
+	"runtime"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+)
+
+// UDPAnyConn extends net.PacketConn to allow reporting the destination IP
+// of incoming packets, and setting the source IP of outgoing packets.  This
+// is relevant for UDP connections that are bound to `0.0.0.0` or `::`.  In
+// these cases, net.PacketConn is not sufficient to enable sending a reply
+// from the expected source IP.
+type UDPAnyConn interface {
+	net.PacketConn
+	ReadToFrom(p []byte) (n int, src *net.UDPAddr, dst net.IP, err error)
```

Comment:
Do we need to couple read and write?

I've been finding a lot cleaner to keep them separate when possible.

See my comment about a Shadowsocks writer below.

Reply:
As noted above, we also need a few other `PacketConn` methods, so it's not as simple as decomposing this into a `Reader` and a `Writer`.